File: SmartShop/CaptureViewController.m
//
//  CaptureViewController.m
//  SmartShop
//
//  Created by Chris on 12/14/14.
//  Copyright (c) 2014 ChrisOtto. All rights reserved.
//

#import <AVFoundation/AVFoundation.h>
#import "CaptureViewController.h"
#import "Item.h"
#import "ItemManager.h"
#import "ItemCommunicator.h"
#import "DBManager.h"
#import "ItemViewController.h"

@interface CaptureViewController () <AVCaptureMetadataOutputObjectsDelegate, ItemManagerDelegate>
{
    AVCaptureSession *_session;
    AVCaptureDevice *_device;
    AVCaptureDeviceInput *_input;
    AVCaptureMetadataOutput *_output;
    AVCaptureVideoPreviewLayer *_prevLayer;

    UIView *_highlightView;
    UILabel *_label;
    
    NSArray *_items;
    ItemManager *_manager;
}

@property (nonatomic, strong) DBManager *dbManager;



@end

@implementation CaptureViewController

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
    
    // capture load
    _highlightView = [[UIView alloc] init];
    _highlightView.autoresizingMask = UIViewAutoresizingFlexibleTopMargin|UIViewAutoresizingFlexibleLeftMargin|UIViewAutoresizingFlexibleRightMargin|UIViewAutoresizingFlexibleBottomMargin;
    _highlightView.layer.borderColor = [UIColor greenColor].CGColor;
    _highlightView.layer.borderWidth = 3;
    [self.view addSubview:_highlightView];
    
    _label = [[UILabel alloc] init];
    _label.frame = CGRectMake(0, self.view.bounds.size.height - 40, self.view.bounds.size.width, 40);
    _label.autoresizingMask = UIViewAutoresizingFlexibleTopMargin;
    _label.backgroundColor = [UIColor colorWithWhite:0.15 alpha:0.65];
    _label.textColor = [UIColor whiteColor];
    _label.textAlignment = NSTextAlignmentCenter;
    _label.text = @"(none)";
    [self.view addSubview:_label];
    
    _session = [[AVCaptureSession alloc] init];
    _device = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
    NSError *error = nil;
    
    _input = [AVCaptureDeviceInput deviceInputWithDevice:_device error:&error];
    if (_input) {
        [_session addInput:_input];
    } else {
        NSLog(@"Error: %@", error);
    }
    
    _output = [[AVCaptureMetadataOutput alloc] init];
    [_output setMetadataObjectsDelegate:self queue:dispatch_get_main_queue()];
    [_session addOutput:_output];
    
    _output.metadataObjectTypes = [_output availableMetadataObjectTypes];
    
    _prevLayer = [AVCaptureVideoPreviewLayer layerWithSession:_session];
    _prevLayer.frame = self.view.bounds;
    _prevLayer.videoGravity = AVLayerVideoGravityResizeAspectFill;
    [self.view.layer addSublayer:_prevLayer];
    
    [_session startRunning];
    
    [self.view bringSubviewToFront:_highlightView];
    [self.view bringSubviewToFront:_label];
    
    //JSON
    _manager = [[ItemManager alloc] init];
    _manager.communicator = [[ItemCommunicator alloc] init];
    _manager.communicator.delegate = _manager;
    _manager.delegate = self;
    
    // Database manager
    // Initialize the dbManager object.
    self.dbManager = [[DBManager alloc] initWithDatabaseFilename:@"smartShop.db"];
}

-(void)startFetchingItem:(NSString *)barcode {
    [_manager fetchItems:barcode];
}

-(void)didReceiveItems:(NSArray *)items {
    _items = items;
}

-(void)fetchingItemsFailedWithError:(NSError *)error {
    NSLog(@"Error %@, %@", error, [error localizedDescription]);
}

- (void)captureOutput:(AVCaptureOutput *)captureOutput didOutputMetadataObjects:(NSArray *)metadataObjects fromConnection:(AVCaptureConnection *)connection
{
    CGRect highlightViewRect = CGRectZero;
    AVMetadataMachineReadableCodeObject *barCodeObject;
    NSString *detectionString = nil;
    NSArray *barCodeTypes = @[AVMetadataObjectTypeUPCECode, AVMetadataObjectTypeCode39Code, AVMetadataObjectTypeCode39Mod43Code,
                              AVMetadataObjectTypeEAN13Code, AVMetadataObjectTypeEAN8Code, AVMetadataObjectTypeCode93Code, AVMetadataObjectTypeCode128Code,
                              AVMetadataObjectTypePDF417Code, AVMetadataObjectTypeQRCode, AVMetadataObjectTypeAztecCode];
    
    for (AVMetadataObject *metadata in metadataObjects) {
        for (NSString *type in barCodeTypes) {
            if ([metadata.type isEqualToString:type])
            {
                barCodeObject = (AVMetadataMachineReadableCodeObject *)[_prevLayer transformedMetadataObjectForMetadataObject:(AVMetadataMachineReadableCodeObject *)metadata];
                highlightViewRect = barCodeObject.bounds;
                detectionString = [(AVMetadataMachineReadableCodeObject *)metadata stringValue];
                break;
            }
        }
        
        if (detectionString != nil)
        {
            _label.text = detectionString;
            [_session stopRunning];
            
            UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Barcode" message:@"Barcode Captured" delegate:self cancelButtonTitle:@"Add" otherButtonTitles:nil];
            alert.alertViewStyle = UIAlertViewStylePlainTextInput;
            UITextField *alertTextField = [alert textFieldAtIndex:0];
            alertTextField.text = detectionString;
            [alert show];
        }
        else
            _label.text = @"(none)";
    }
    
    _highlightView.frame = highlightViewRect;
}

- (void)alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)buttonIndex{
    NSString *barcode = [[alertView textFieldAtIndex:0] text];
    [self startFetchingItem:barcode];
    
    Item *item = _items;
    NSString *name = item.name;
    NSString *attributes = item.attributes;
    
    NSString *query = [NSString stringWithFormat:@"insert into items values (null, %d, '%@', '%@', null, null, '%@')", _recordIDToView, barcode, name, attributes];
    
    // Execute the query.
    [self.dbManager executeQuery:query];
    
    // If the query was successfully executed then pop the view controller.
    if (self.dbManager.affectedRows != 0) {
        NSLog(@"Query was executed successfully. Affected rows = %d", self.dbManager.affectedRows);
        
        // Pop the view controller.
        [self.navigationController popViewControllerAnimated:YES];
    }
    else{
        NSLog(@"Could not execute the query.");
    }

}

@end

File: SmartShop/CaptureViewController.swift
//  Converted to Swift 4 by Swiftify v4.2.20229 - https://objectivec2swift.com/
//
//  CaptureViewController.swift
//  SmartShop
//
//  Created by Chris on 12/14/14.
//  Copyright (c) 2014 ChrisOtto. All rights reserved.
//

import AVFoundation
import UIKit

class CaptureViewController: UIViewController, AVCaptureMetadataOutputObjectsDelegate, ItemManagerDelegate {
    private var session: AVCaptureSession?
    private var device: AVCaptureDevice?
    private var input: AVCaptureDeviceInput?
    private var output: AVCaptureMetadataOutput?
    private var prevLayer: AVCaptureVideoPreviewLayer?
    private var highlightView: UIView?
    private var label: UILabel?
    private var items: [Any] = []
    private var manager: ItemManager?

    var recordIDToView: Int = 0
    private var dbManager: DBManager?

    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        
        // Custom initialization
    
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.

        // capture load
        highlightView = UIView()
        highlightView?.autoresizingMask = [.flexibleTopMargin, .flexibleLeftMargin, .flexibleRightMargin, .flexibleBottomMargin]
        highlightView?.layer.borderColor = UIColor.green.cgColor
        highlightView?.layer.borderWidth = 3
        if let highlightView = highlightView {
            view.addSubview(highlightView)
        }

        label = UILabel()
        label?.frame = CGRect(x: 0, y: view.bounds.size.height - 40, width: view.bounds.size.width, height: 40)
        label?.autoresizingMask = .flexibleTopMargin
        label?.backgroundColor = UIColor(white: 0.15, alpha: 0.65)
        label?.textColor = UIColor.white
        label?.textAlignment = .center
        label?.text = "(none)"
        if let label = label {
            view.addSubview(label)
        }

        session = AVCaptureSession()
        device = AVCaptureDevice.default(for: .video)
        var error: Error? = nil

        if let device = device {
            input = try? AVCaptureDeviceInput(device: device)
        }
        if input != nil {
            if let input = input {
                session?.addInput(input)
            }
        } else {
            if let error = error {
                print("Error: \(error)")
            }
        }

        output = AVCaptureMetadataOutput()
        output?.setMetadataObjectsDelegate(self, queue: DispatchQueue.main)
        if let output = output {
            session?.addOutput(output)
        }

        if let available = output?.availableMetadataObjectTypes {
            output?.metadataObjectTypes = available
        }

        if let session = session {
            prevLayer = AVCaptureVideoPreviewLayer(session: session)
        }
        prevLayer?.frame = view.bounds
        prevLayer?.videoGravity = .resizeAspectFill
        if let prevLayer = prevLayer {
            view.layer.addSublayer(prevLayer)
        }

        session?.startRunning()

        if let highlightView = highlightView {
            view.bringSubviewToFront(highlightView)
        }
        if let label = label {
            view.bringSubviewToFront(label)
        }

        //JSON
        manager = ItemManager()
        manager?.communicator = ItemCommunicator()
        manager?.communicator?.delegate = manager
        manager?.delegate = self

        // Database manager
        // Initialize the dbManager object.
        dbManager = DBManager(databaseFilename: "smartShop.db")
    }

    func startFetchingItem(_ barcode: String?) {
        manager?.fetchItems(barcode)
    }

    func didReceiveItems(_ items: [Any]?) {
        self.items = items
    }

    func fetchingItemsFailed() throws {
        if let error = error {
            print("Error \(error), \(error?.localizedDescription ?? "")")
        }
    }

    func metadataOutput(_ captureOutput: AVCaptureMetadataOutput, didOutput metadataObjects: [AVMetadataObject], from connection: AVCaptureConnection) {
        var highlightViewRect = CGRect.zero
        var barCodeObject: AVMetadataMachineReadableCodeObject?
        var detectionString: String? = nil
        let barCodeTypes = [
            .upce,
            .code39,
            .code39Mod43,
            .ean13,
            .ean8,
            .code93,
            .code128,
            .pdf417,
            .qr,
            .aztec
        ] as? [Any]

        for metadata: AVMetadataObject in metadataObjects {
            for type: String? in barCodeTypes as? [String?] ?? [] {
                if (metadata.type == type) {
                    barCodeObject = prevLayer?.transformedMetadataObject(for: metadata as? AVMetadataMachineReadableCodeObject) as? AVMetadataMachineReadableCodeObject
                    highlightViewRect = barCodeObject?.bounds ?? CGRect.zero
                    detectionString = (metadata as? AVMetadataMachineReadableCodeObject)?.stringValue
                    break
                }
            }

            if detectionString != nil {
                label?.text = detectionString
                session?.stopRunning()

                let alert = UIAlertView(title: "Barcode", message: "Barcode Captured", delegate: self, cancelButtonTitle: "Add", otherButtonTitles: "")
                alert.alertViewStyle = .plainTextInput
                let alertTextField: UITextField? = alert.textField(at: 0)
                alertTextField?.text = detectionString
                alert.show()
            } else {
                label?.text = "(none)"
            }
        }

        highlightView?.frame = highlightViewRect
    }

    func alertView(_ alertView: UIAlertView, clickedButtonAt buttonIndex: Int) {
        let barcode = alertView.textField(at: 0)?.text
        startFetchingItem(barcode)

        let item = items as? Item
        let name = item?.name
        let attributes = item?.attributes

        let query = "insert into items values (null, \(recordIDToView), '\(barcode ?? "")', '\(name ?? "")', null, null, '\(attributes ?? "")')"

        // Execute the query.
        dbManager?.executeQuery(query)

        // If the query was successfully executed then pop the view controller.
        if dbManager?.affectedRows != 0 {
            print("Query was executed successfully. Affected rows = \(dbManager?.affectedRows ?? 0)")

            // Pop the view controller.
            navigationController?.popViewController(animated: true)
        } else {
            print("Could not execute the query.")
        }

    }

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
}